```
src/peer_connection.cpp:388:15: error: declaration shadows a local variable [-Werror,-Wshadow]
                        error_code ec;
                                   ^
src/peer_connection.cpp:377:14: note: previous declaration is here
                error_code ec;
                           ^
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -7,6 +7,7 @@
 	* improve dont-count-slow-torrents
 	* improve performance of checking files
 	* add ip_ban_alert for IPs banned internally by libtorrent
+	* fix peer DSCP marking on outgoing TCP connections
 
 2.1.0 release
 

--- a/simulation/test_outgoing_interfaces.cpp
+++ b/simulation/test_outgoing_interfaces.cpp
@@ -29,8 +29,8 @@ namespace {
 	// src/enum_net.cpp).
 	constexpr char const* device_name = "eth0";
 
-	template <typename Settings, typename Test>
-	void run_test(Settings const& make_settings, Test const& test)
+	template <typename Settings, typename Alert, typename Test>
+	void run_test(Settings const& make_settings, Alert const& on_alert, Test const& test)
 	{
 		// setup the simulation
 		sim::default_config network_cfg;
@@ -56,12 +56,13 @@ namespace {
 		params.flags &= ~lt::torrent_flags::paused;
 		ses->async_add_torrent(params);
 
-		print_alerts(*ses, [](lt::session&, lt::alert const* a) {
+		print_alerts(*ses, [&](lt::session&, lt::alert const* a) {
 			if (auto at = lt::alert_cast<lt::add_torrent_alert>(a))
 			{
 				lt::torrent_handle h = at->handle;
 				add_fake_peer(h, 0);
 			}
+			on_alert(a);
 		});
 
 		sim::timer t(sim, lt::seconds(60), [&](boost::system::error_code const&) {
@@ -74,6 +75,12 @@ namespace {
 		});
 
 		sim.run();
+	}
+
+	template <typename Settings, typename Test>
+	void run_test(Settings const& make_settings, Test const& test)
+	{
+		run_test(make_settings, [](lt::alert const*) {}, test);
 	}
 
 } // anonymous namespace
@@ -123,3 +130,37 @@ TORRENT_TEST(outgoing_interfaces_matching_device_allowed_with_proxy)
 		},
 		[](fake_peer& p) { TEST_EQUAL(p.connected(), true); });
 }
+
+#ifndef TORRENT_DISABLE_LOGGING
+// peer_dscp must be applied after opening the socket but before initiating
+// the outgoing TCP connection. Otherwise set_traffic_class() fails with
+// error::bad_descriptor and the SYN is sent without the configured DSCP.
+TORRENT_TEST(outgoing_peer_dscp)
+{
+	bool socket_opened = false;
+	bool dscp_error = false;
+	run_test(
+		[](lt::settings_pack& pack) {
+			pack.set_int(lt::settings_pack::peer_dscp, 0x04);
+			pack.set_bool(lt::settings_pack::enable_outgoing_utp, false);
+		},
+		[&](lt::alert const* a) {
+			auto const* pl = lt::alert_cast<lt::peer_log_alert>(a);
+			if (pl == nullptr)
+				return;
+
+			if (pl->event_type == lt::peer_log_alert::open)
+				socket_opened = true;
+			else if (pl->event_type == lt::peer_log_alert::set_dscp)
+			{
+				TEST_EQUAL(socket_opened, true);
+				dscp_error = true;
+			}
+		},
+		[&](fake_peer& p) {
+			TEST_EQUAL(p.connected(), true);
+			TEST_EQUAL(socket_opened, true);
+			TEST_EQUAL(dscp_error, false);
+		});
+}
+#endif

--- a/src/peer_connection.cpp
+++ b/src/peer_connection.cpp
@@ -309,10 +309,13 @@ namespace {
 				int const value = m_settings.get_int(settings_pack::peer_dscp);
 				aux::set_traffic_class(m_socket, value, ec);
 #ifndef TORRENT_DISABLE_LOGGING
-				if (ec && should_log(peer_log_alert::outgoing))
+				if (ec && should_log(peer_log_alert::incoming))
 				{
-					peer_log(peer_log_alert::outgoing, peer_log_alert::set_dscp, "value: %d e: %s"
-						, value, ec.message().c_str());
+					peer_log(peer_log_alert::incoming,
+						peer_log_alert::set_dscp,
+						"value: %d e: %s",
+						value,
+						ec.message().c_str());
 				}
 #endif
 			}
@@ -347,20 +350,6 @@ namespace {
 			init();
 		}
 
-		if (m_settings.get_int(settings_pack::peer_dscp) != 0)
-		{
-			int const value = m_settings.get_int(settings_pack::peer_dscp);
-			error_code ec;
-			aux::set_traffic_class(m_socket, value, ec);
-#ifndef TORRENT_DISABLE_LOGGING
-			if (ec && should_log(peer_log_alert::outgoing))
-			{
-				peer_log(peer_log_alert::outgoing, peer_log_alert::set_dscp, "value: %d e: %s"
-					, value, ec.message().c_str());
-			}
-#endif
-		}
-
 		// if this is an incoming connection, we're done here
 		if (!m_connecting)
 		{
@@ -391,6 +380,23 @@ namespace {
 		{
 			disconnect(ec, operation_t::sock_open);
 			return;
+		}
+
+		if (m_settings.get_int(settings_pack::peer_dscp) != 0)
+		{
+			int const value = m_settings.get_int(settings_pack::peer_dscp);
+			error_code err;
+			aux::set_traffic_class(m_socket, value, err);
+#ifndef TORRENT_DISABLE_LOGGING
+			if (err && should_log(peer_log_alert::outgoing))
+			{
+				peer_log(peer_log_alert::outgoing,
+					peer_log_alert::set_dscp,
+					"value: %d e: %s",
+					value,
+					err.message().c_str());
+			}
+#endif
 		}
 
 		tcp::endpoint const bound_ip = m_ses.bind_outgoing_socket(m_socket

